### PR TITLE
Employee can add issues

### PR DESCRIPTION
**Pull request by paromita92**

### CODE REVIEW

File: app/controllers/employee/issues_controller.rb [hunk 1]
```diff
@@ -0,0 +1,67 @@
+class Employee::IssuesController < ApplicationController
+  def index
+    @issues = Issue.includes(:resolution, item: [:brand, :category]).where(items: { employee_id: current_user.employee.id })
```
Comment (nbanerjee) [inline app/controllers/employee/issues_controller.rb]: use association for this

File: app/controllers/employee/issues_controller.rb [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+
+    if params[:item_id].present?
+      @item = Item.find(params[:item_id])
+      @issues = @issues.where(item_id: @item.id, items: { employee_id: current_user.employee.id })
+    end
```
Comment (nbanerjee) [inline app/controllers/employee/issues_controller.rb]: Why you need this ?

File: app/controllers/employee/issues_controller.rb [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+  def edit
+    @issue = current_user.items.issues.find(params[:id])
+  end
+
+  def show
+    @issue = current_user.items.issues.find(params[:id])
+  end
+
+  def create
+    @issue = Issue.new(issue_params)
```
Comment (nbanerjee) [inline app/controllers/employee/issues_controller.rb]: use employee reference at the time of creating issue

File: app/controllers/employee/issues_controller.rb [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+
+    if @issue.save
+      redirect_to :back, flash: { success: t('create') } unless request.xhr?
+    else
+      render :new
+    end
+  end
+
+  def update
+    @issue = Issue.find(params[:id])
```
Comment (nbanerjee) [inline app/controllers/employee/issues_controller.rb]: add access control here.

File: app/controllers/employee/issues_controller.rb [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+
+    if @issue.update_attributes(issue_params)
+      redirect_to :back, flash: { success: t('update') } unless request.xhr?
+    else
+      render :edit
+    end
+  end
+
+  def set_priority
+    @issue = Issue.find(params[:id])
```
Comment (nbanerjee) [inline app/controllers/employee/issues_controller.rb]: add access control here.

File: app/helpers/items_helper.rb [hunk 6]
```diff
@@ -12,4 +12,12 @@ def checkout_details(item)
...
     value
   end
+
+  def admin_items_list
+    options_for_select(Item.not_erased.active.available.sort_by(&:name).map{ |i| [i.name_with_id, i.id, { 'data-purchase-on' => i.purchase_on }] })
+  end
+
+  def employee_items_list
+    options_for_select(current_user.employee.items.not_erased.active.sort_by(&:name).map{ |i| [i.name_with_id, i.id, { 'data-purchase-on' => i.purchase_on }] })
+  end
```
Comment (nbanerjee) [inline app/helpers/items_helper.rb]: No need to create helper.

File: app/controllers/employee/issues_controller.rb [hunk 7]
```diff
@@ -0,0 +1,61 @@
+class Employee::IssuesController < ApplicationController
+  def index
+    @issues = current_user.employee.issues.all
```
Comment (nbanerjee) [inline app/controllers/employee/issues_controller.rb]: remove `.all`, check n+1 query

File: app/controllers/employee/issues_controller.rb [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+    @issues = @issues.paginate(page: params[:page])
+  end
+
+  def new
+    if item = Item.find_by_id(params[:item_id])
```
Comment (nbanerjee) [inline app/controllers/employee/issues_controller.rb]: add access control.

**Comment by nbanerjee** [comment]
Please remove user model. And incorporate all user level features into employee.

File: app/views/checkouts/_form.html.erb [hunk 9]
```diff
@@ -1,6 +1,12 @@
 <div class="form-group row">
   <div class="col-sm-4"><%= f.label :item_id , t('.title.item'), class: 'required' %></div>
-  <div class="col-sm-8"><%= f.select :item_id, options_for_select(Item.not_erased.active.available.sort_by(&:name).map{ |i| [i.name_with_id, i.id, { 'data-purchase-on' => i.purchase_on }] }), { prompt: t('.placeholder.select_item') }, class: 'form-control' %></div>
+  <div class="col-sm-8">
+    <% if current_employee.admin %>
+      <%= f.select :item_id, admin_items_list, { prompt: t('.placeholder.select_item') }, class: 'form-control' %>
+    <% else %>
+      <%= f.select :item_id, employee_items_list, { prompt: t('.placeholder.select_item') }, class: 'form-control' %>
+    <% end %>
+  </div>
```
Comment (nbanerjee) [inline app/views/checkouts/_form.html.erb]: This change should not be here.

File: app/controllers/employee/issues_controller.rb [hunk 10]
```diff
@@ -0,0 +1,62 @@
...
+
+  def new
+    if item = current_employee.items.find_by_id(params[:item_id])
+      @issue = item.issues.build
+    else
+      @issue = Issue.new
+    end
+  end
+
+
```
Comment (nbanerjee) [inline app/controllers/employee/issues_controller.rb]: Remove unwanted blank line.

File: app/controllers/employee/issues_controller.rb [hunk 11]
```diff
@@ -0,0 +1,62 @@
...
+  def edit
+    @issue = current_employee.issues.find(params[:id])
+  end
+
+  def show
+    @issue = current_employee.issues.find(params[:id])
+  end
+
+  def create
+    @issue = current_employee.issues.new(issue_params)
```
Comment (nbanerjee) [inline app/controllers/employee/issues_controller.rb]: use build

File: app/models/employee.rb [hunk 12]
```diff
@@ -1,17 +1,35 @@
...
   scope :order_by_name, -> { order('name') }
 
-  def deallocate_items
-    items.clear() if active_changed? && active == false
+  def deactivate_employee_and_deallocate_items
+    if active_changed? && active == false
+      items.clear()
+      employee.update(active: false) if employee.present?
+    end
+  end
```
Comment (nbanerjee) [inline app/models/employee.rb]: Why have you changed this code?

File: app/views/layouts/_navigation.html.erb [hunk 13]
```diff
@@ -11,47 +11,52 @@
...
+              <%= link_to t('.list.issues'), employee_issues_path %>
+            </li>
+	   <% end %>
         </ul>
         <ul class="nav navbar-nav navbar-right">
           <li class="dropdown menu-custom">
             <a href="#" class="dropdown-toggle" data-toggle="dropdown">
               <i class="fa fa-user"></i>&nbsp;
-              <%= current_user.full_name.present? ? current_user.full_name : current_user.email %> <b class="caret"></b>
+              <%= current_employee.name.present? ? current_employee.name : current_employee.email %> <b class="caret"></b>
```
Comment (nbanerjee) [inline app/views/layouts/_navigation.html.erb]: Please add a model method and use that

File: db/migrate/20170222061836_add_admin_to_users.rb [hunk 14]
```diff
@@ -0,0 +1,6 @@
+class AddAdminToUsers < ActiveRecord::Migration
+  def change
+    add_column :users, :admin, :boolean, default: false, null: false
+    add_column :users, :active, :boolean, default: true, null: false
+  end
+end
```
Comment (nbanerjee) [inline db/migrate/20170222061836_add_admin_to_users.rb]: Remove unwanted migrations

File: db/migrate/20170309072540_add_user_id_to_employees.rb [hunk 15]
```diff
@@ -0,0 +1,5 @@
+class AddUserIdToEmployees < ActiveRecord::Migration
+  def change
+    add_reference :employees, :user, index: true, foreign_key: true
+  end
+end
```
Comment (nbanerjee) [inline db/migrate/20170309072540_add_user_id_to_employees.rb]: Remove unwanted migrations

File: db/migrate/20170310114118_add_user_id_to_existing_employees.rb [hunk 16]
```diff
@@ -0,0 +1,14 @@
...
+    employees.each do |emp|
+      user = User.find_or_create_by(email: emp.email)
+      emp.user_id = user.id
+      emp.save
+    end
+  end
+
+  def down
+  end
+end
```
Comment (nbanerjee) [inline db/migrate/20170310114118_add_user_id_to_existing_employees.rb]: Remove unwanted migrations

File: db/migrate/20170420093356_drop_table.rb [hunk 17]
```diff
@@ -0,0 +1,8 @@
+class DropTable < ActiveRecord::Migration
+  def up
+    drop_table :users
```
Comment (nbanerjee) [inline db/migrate/20170420093356_drop_table.rb]: No need to create different migration

File: app/models/employee.rb [hunk 18]
```diff
@@ -1,17 +1,38 @@
...
 
   scope :order_by_name, -> { order('name') }
 
-  def deallocate_items
-    items.clear() if active_changed? && active == false
+  def self.employee_present(current_employee)
+    current_employee.name.present? ? current_employee.name : current_employee.email
+  end
+
+  def deactivate_employee_and_deallocate_items
```
Comment (nbanerjee) [inline app/models/employee.rb]: Why you need to change this name? revert back this method.

File: app/views/issues/_form.html.erb [hunk 19]
```diff
@@ -5,15 +5,22 @@
...
+  <% if current_employee.admin %>
+    <div class="col-sm-4"><%= f.label :employee %></div>
+    <div class="col-sm-8">
+      <%= f.collection_select :employee_id, Employee.all, :id, :name, { include_blank: t('.placeholder.select') }, class: 'form-control' %>
+    </div>
+  <% end %>
+ </div>
 <div class="form-group row">
   <div class="col-sm-4"><%= f.label :description, t('.title.description') %></div>
   <div class="col-sm-8"><%= f.text_area :description, class: 'form-control', placeholder: t('.placeholder.description') %></div>
```
Comment (nbanerjee) [inline app/views/issues/_form.html.erb]: Use this form in employee/issue

File: app/views/layouts/_navigation.html.erb [hunk 20]
```diff
@@ -11,47 +11,52 @@
...
+              <%= link_to t('.list.issues'), employee_issues_path %>
+            </li>
+	   <% end %>
         </ul>
         <ul class="nav navbar-nav navbar-right">
           <li class="dropdown menu-custom">
             <a href="#" class="dropdown-toggle" data-toggle="dropdown">
               <i class="fa fa-user"></i>&nbsp;
-              <%= current_user.full_name.present? ? current_user.full_name : current_user.email %> <b class="caret"></b>
+              <%= Employee.employee_present(current_employee) %> <b class="caret"></b>
```
Comment (nbanerjee) [inline app/views/layouts/_navigation.html.erb]: What is this? What you want to do here?

File: app/controllers/employee/issues_controller.rb [hunk 21]
```diff
@@ -0,0 +1,61 @@
...
+    end
+  end
+
+  private
+
+  def issue_params
+    params.require(:issue).permit(:item_id, :title, :description, :note, :priority)
+  end
+
+  def resolution_params
```
Comment (nbanerjee) [inline app/controllers/employee/issues_controller.rb]: Where have you used this method?

File: app/models/employee.rb [hunk 22]
```diff
@@ -1,17 +1,40 @@
...
+  validates :email, uniqueness: { case_sensitive: false }, format: { with: VALID_EMAIL_REGEX }, allow_blank: true
 
-  before_save :deallocate_items
+  before_save  :deactivate_employee_and_deallocate_items
 
   scope :order_by_name, -> { order('name') }
 
-  def deallocate_items
-    items.clear() if active_changed? && active == false
+  def employee_present(current_employee)
```
Comment (nbanerjee) [inline app/models/employee.rb]: Why method name and its definition is completely different? rename it `name_or_email` and remove unwanted ugly code.

File: app/models/employee.rb [hunk 23]
```diff
@@ -1,17 +1,40 @@
...
   has_many :item_histories, dependent: :nullify
   has_many :checkouts, dependent: :nullify
   has_many :issues, dependent: :nullify
 
   validates :name, presence: true
-  validates :email, uniqueness: { case_sensitive: false }, format: { with: User::VALID_EMAIL_REGEX }, allow_blank: true
+  validates :email, uniqueness: { case_sensitive: false }, format: { with: VALID_EMAIL_REGEX }, allow_blank: true
 
-  before_save :deallocate_items
+  before_save  :deactivate_employee_and_deallocate_items
```
Comment (nbanerjee) [inline app/models/employee.rb]: No need to change here.

File: app/views/layouts/_navigation.html.erb [hunk 24]
```diff
@@ -11,47 +11,52 @@
...
+              <%= link_to t('.list.issues'), employee_issues_path %>
+            </li>
+	   <% end %>
         </ul>
         <ul class="nav navbar-nav navbar-right">
           <li class="dropdown menu-custom">
             <a href="#" class="dropdown-toggle" data-toggle="dropdown">
               <i class="fa fa-user"></i>&nbsp;
-              <%= current_user.full_name.present? ? current_user.full_name : current_user.email %> <b class="caret"></b>
+              <%= current_employee.name_or_email(current_employee) %> <b class="caret"></b>
```
Comment (nbanerjee) [inline app/views/layouts/_navigation.html.erb]: Why you need to send `current_employee` in param of an instance method?

File: app/views/issues/_form.html.erb [hunk 25]
```diff
@@ -5,7 +5,7 @@
 <div class="form-group row">
   <div class="col-sm-4"><%= f.label :items, t('.title.item') %></div>
   <div class="col-sm-8">
-    <%= f.select :item_id, options_for_select(Item.not_erased.active.sort_by(&:name).map{ |i| [i.name_with_id, i.id, { 'data-purchase-on' => i.purchase_on }] }, selected: @issue.item_id), { include_blank: t('.placeholder.select') }, class: 'form-control' %>
+   <%= f.select :item_id, options_for_select(Item.not_erased.active.sort_by(&:name).map{ |i| [i.name_with_id, i.id, { 'data-purchase-on' => i.purchase_on }] }, selected: @issue.item_id), { include_blank: t('.placeholder.select') }, class: 'form-control' %>
```
Comment (nbanerjee) [inline app/views/issues/_form.html.erb]: No change here only you have removed correct indentation !

File: config/routes.rb [hunk 26]
```diff
@@ -47,9 +47,18 @@
...
   delete 'logout', to: 'sessions#destroy'
+
+  namespace :employee do
+    resources :issues, only: [:edit, :update, :new, :create, :index] do
+      member do
+        put 'set_priority'
+      end
+    end
+
+    resources :checkouts, except: [:destroy, :checkin]
```
Comment (nbanerjee) [inline config/routes.rb]: this is an out of context change, remove it.

File: db/migrate/20170420080431_update_employees.rb [hunk 27]
```diff
@@ -0,0 +1,16 @@
+class UpdateEmployees < ActiveRecord::Migration
+  def self.up
```
Comment (nbanerjee) [inline db/migrate/20170420080431_update_employees.rb]: no need to write `self.`

File: app/models/employee.rb [hunk 28]
```diff
@@ -1,17 +1,40 @@
...
+  validates :email, uniqueness: { case_sensitive: false }, format: { with: VALID_EMAIL_REGEX }, allow_blank: true
 
-  before_save :deallocate_items
+  before_save  :deallocate_items
 
   scope :order_by_name, -> { order('name') }
 
+  def name_or_email
+    name
+    email
```
Comment (nbanerjee) [inline app/models/employee.rb]: What you want to do here?

File: app/models/employee.rb [hunk 29]
```diff
@@ -1,17 +1,40 @@
...
+    name.present? ? name : email
```
Comment (nbanerjee) [inline app/models/employee.rb]: `name.presence || email`

File: app/models/employee.rb [hunk 30]
```diff
@@ -1,17 +1,40 @@
...
   has_many :item_histories, dependent: :nullify
   has_many :checkouts, dependent: :nullify
   has_many :issues, dependent: :nullify
 
   validates :name, presence: true
-  validates :email, uniqueness: { case_sensitive: false }, format: { with: User::VALID_EMAIL_REGEX }, allow_blank: true
+  validates :email, uniqueness: { case_sensitive: false }, format: { with: VALID_EMAIL_REGEX }, allow_blank: true
 
-  before_save :deallocate_items
+  before_save  :deallocate_items
```
Comment (nbanerjee) [inline app/models/employee.rb]: remove unwanted space

File: app/views/employee/issues/_form.html.erb [hunk 31]
```diff
@@ -0,0 +1,32 @@
...
+ </div>
+<div class="form-group row">
+  <div class="col-sm-4"><%= f.label :description, t('.title.description') %></div>
+  <div class="col-sm-8"><%= f.text_area :description, class: 'form-control', placeholder: t('.placeholder.description') %></div>
+</div>
+<div class="form-group row">
+  <div class="col-sm-4"><%= f.label :priority, t('.title.priority') %></div>
+  <div class="col-sm-8"><%= f.select :priority, Issue.priorities.sort_by{ |key, value| key }.map{ |key, value| [key.humanize, key] }, { include_blank: t('.placeholder.select') }, class: 'form-control' %></div>
+</div>
+<%= unobtrusive_js("$('.datepicker').datetimepicker({ format: 'DD/MM/YYYY' });") %>
```
Comment (nbanerjee) [inline app/views/employee/issues/_form.html.erb]: Please use existing form

File: app/views/issues/_form.html.erb [hunk 32]
```diff
@@ -5,7 +5,7 @@
 <div class="form-group row">
   <div class="col-sm-4"><%= f.label :items, t('.title.item') %></div>
   <div class="col-sm-8">
-    <%= f.select :item_id, options_for_select(Item.not_erased.active.sort_by(&:name).map{ |i| [i.name_with_id, i.id, { 'data-purchase-on' => i.purchase_on }] }, selected: @issue.item_id), { include_blank: t('.placeholder.select') }, class: 'form-control' %>
+     <%= f.select :item_id, options_for_select(Item.not_erased.active.sort_by(&:name).map{ |i| [i.name_with_id, i.id, { 'data-purchase-on' => i.purchase_on }] }, selected: @issue.item_id), { include_blank: t('.placeholder.select') }, class: 'form-control' %>
```
Comment (nbanerjee) [inline app/views/issues/_form.html.erb]: indentation!!

File: app/views/issues/_form.html.erb [hunk 33]
```diff
@@ -8,12 +8,15 @@
...
   </div>
 </div>
+<% if current_employee.admin %>
 <div class="form-group row">
   <div class="col-sm-4"><%= f.label :employee %></div>
   <div class="col-sm-8">
     <%= f.collection_select :employee_id, Employee.all, :id, :name, { include_blank: t('.placeholder.select') }, class: 'form-control' %>
   </div>
 </div>
+<% end %>
```
Comment (nbanerjee) [inline app/views/issues/_form.html.erb]: Indentation !!

File: app/views/employee/issues/_form.html.erb [hunk 34]
```diff
@@ -0,0 +1 @@
+<%= render partial: 'issues/form', locals: {f:f} %>
```
Comment (nbanerjee) [inline app/views/employee/issues/_form.html.erb]: remove partial and locals from here, use new syntax. Better directly use it.

File: app/views/employee/issues/new.html.erb [hunk 35]
```diff
@@ -0,0 +1,8 @@
+<%= form_for @issue, url: employee_issues_path, remote: true, html: { class: "form-horizontal" } do |f| %>
+  <%= render 'form', f:f %>
+  <div class="text-right">
+    <button type="button" class="btn btn-default" data-dismiss="modal"><%= t('button.cancel') %></button>
+    <%= f.submit t('button.create'), class: "btn btn-primary" %>
+  </div>
+<% end %>
+
```
Comment (nbanerjee) [inline app/views/employee/issues/new.html.erb]: remove extra line

File: app/views/employees/index.html.erb [hunk 36]
```diff
@@ -23,10 +23,9 @@
 	   <% end %>
 	 </td>
         <td>
-          <% unless employee.external_id? %>
             <%= edit_link(edit_employee_path(employee), t('.edit.popup.title')) %>
-            <%= delete_link(employee_path(employee))  %>
-          <% end %>  
+            <%= delete_link(employee_path(employee)) unless employee == current_employee %>
+
```
Comment (nbanerjee) [inline app/views/employees/index.html.erb]: remove unwanted lines

File: app/models/employee.rb [hunk 37]
```diff
@@ -1,17 +1,38 @@
...
+
   def deallocate_items
-    items.clear() if active_changed? && active == false
+    if active_changed? && active == false
+      items.clear()
+    end
+  end
+
+  def self.find_and_update_from_auth(auth)
+    employee ||= Employee.find_by_email(auth.info.email)
```
Comment (nbanerjee) [inline app/models/employee.rb]: use new syntax `Employee.find_by(email: ...)`

File: app/views/employee/issues/edit.html.erb [hunk 38]
```diff
@@ -0,0 +1,7 @@
+<%= form_for @issue, url: employee_issue_path(@issue), method: :patch, remote: true, html: { class: "form-horizontal" } do |f| %>
+  <%= render 'form', f:f %>
```
Comment (nbanerjee) [inline app/views/employee/issues/edit.html.erb]: add space after `f: `

File: app/views/employees/index.html.erb [hunk 39]
```diff
@@ -23,10 +23,8 @@
 	   <% end %>
 	 </td>
         <td>
-          <% unless employee.external_id? %>
             <%= edit_link(edit_employee_path(employee), t('.edit.popup.title')) %>
-            <%= delete_link(employee_path(employee))  %>
-          <% end %>  
+            <%= delete_link(employee_path(employee)) unless employee == current_employee %>
```
Comment (nbanerjee) [inline app/views/employees/index.html.erb]: Indentation?

**Comment by ssinghi** [comment]
@nbanerjee @paromita92 do you understand what are you doing in this PR? Have you reviewed it? Have you tested it?

File: db/schema.rb [hunk 40]
```diff
@@ -57,12 +57,15 @@
...
 
   create_table "employees", force: :cascade do |t|
-    t.string   "name",        limit: 255,                null: false
-    t.string   "email",       limit: 255
-    t.boolean  "active",                  default: true
-    t.integer  "external_id", limit: 4
-    t.datetime "created_at",             null: false
-    t.datetime "updated_at",             null: false
+    t.string   "name",         limit: 255,                 null: false
+    t.string   "email",        limit: 255
```
Comment (nbanerjee) [inline db/schema.rb]: Email should be `null: false`

File: db/schema.rb [hunk 41]
```diff
@@ -57,12 +57,15 @@
...
+    t.boolean  "active",                   default: true
+    t.integer  "external_id",  limit: 4
+    t.datetime "created_at"
+    t.datetime "updated_at"
+    t.string   "access_token", limit: 255
+    t.string   "google_uid",   limit: 255
+    t.boolean  "admin",                    default: false
   end
```
Comment (nbanerjee) [inline db/schema.rb]: Add unique index for employee, check and create employees corresponding all users.

**Pull request closed without merging**